### ISSUE TITLE
Test and possibly improve shrinking

### ISSUE THREAD

**Issue by effectfully** [effectfully/tiny-lang#8]
I was getting some test failures in a library that depends on this one, but `QuickCheck` was giving me quite non-minimal expressions that break tests. We should investigate that. In particular, the problematic instance seems to be the `Arbitrary (ExprWithEnv f)` one.

**Comment by effectfully** [comment]
Here is one test failure:

```
      *** Failed! (after 52 tests and 2 shrinks):
      ExprWithEnv (SomeUniExpr Bool (EAppBinOp FGe (EIf (EVal False) (EVal 4) (EVal 8)) (EAppBinOp Sub (EVal 2) (EVal 2)))) (Env {unEnv = fromList []})
```

That term is really far from being minimal as the thing is supposed to fail on `SomeUniExpr Bool (EAppBinOp FGe (EVal 0) (EVal 0))`. I.e. when shrinking an applied operator we should also try to shrink its arguments, but do not shrink the operator itself (but this does sound exponentially for binary operators).

... which also means that we need type-preserving shrinking.

**Comment by effectfully** [comment]
Implemented type-preserving shrinking in 9418f285bd57d0998de6d5fec0fa5140aa254804. Now shrinking works like a charm and actually gives minimal counterexamples.
